src/Nethermind/Nethermind.Blockchain/Synchronization/FastSync/NodeDataDownloader.cs

```diff
@@ -22,125 +22,69 @@
 using Nethermind.Core;
 using Nethermind.Core.Crypto;
 using Nethermind.Logging;
+using Nethermind.Store;
 
 namespace Nethermind.Blockchain.Synchronization.FastSync
 {
     public class NodeDataDownloader : INodeDataDownloader
     {
         private readonly IEthSyncPeerPool _syncPeerPool;
-        private readonly INodeDataFeed _nodeDataFeed;
+        private readonly INodeDataFeed _feed;
+        private readonly INodeDataConsumer _additionalConsumer;
         private int _pendingRequests;
         private int _consumedNodesCount;
         private ILogger _logger;
 
-        public NodeDataDownloader(IEthSyncPeerPool syncPeerPool, INodeDataFeed nodeDataFeed, ILogManager logManager)
+        public NodeDataDownloader(IEthSyncPeerPool syncPeerPool, INodeDataFeed nodeDataFeed, INodeDataConsumer additionalConsumer, ILogManager logManager)
         {
             _syncPeerPool = syncPeerPool ?? throw new ArgumentNullException(nameof(syncPeerPool));
-            _nodeDataFeed = nodeDataFeed ?? throw new ArgumentNullException(nameof(nodeDataFeed));
+            _feed = nodeDataFeed ?? throw new ArgumentNullException(nameof(nodeDataFeed));
+            _additionalConsumer = additionalConsumer ?? throw new ArgumentNullException(nameof(additionalConsumer));
             _logger = logManager.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
         }
 
-        private SemaphoreSlim _semaphore = new SemaphoreSlim(0);
-        
-        private int _lastUsefulPeerCount;
-        private int _noAllocInARow;
-        private bool _noAllocPunishment;
-
         private async Task ExecuteRequest(CancellationToken token, StateSyncBatch batch)
         {
-            SyncPeerAllocation nodeSyncAllocation = _syncPeerPool.Borrow(BorrowOptions.DoNotReplace, "node sync");
-            try
+            var peer = batch.AssignedPeer.Current.SyncPeer;
+            if (peer != null)
             {
-                ISyncPeer peer = nodeSyncAllocation?.Current?.SyncPeer;
-                batch.AssignedPeer = nodeSyncAllocation;
-                if (peer != null)
-                {
-                    var hashes = batch.RequestedNodes.Select(r => r.Hash).ToArray();
-                    Task<byte[][]> getNodeDataTask = peer.GetNodeData(hashes, token);
-                    await getNodeDataTask.ContinueWith(
-                        t =>
+                var hashes = batch.RequestedNodes.Select(r => r.Hash).ToArray();
+                Task<byte[][]> getNodeDataTask = peer.GetNodeData(hashes, token);
+                await getNodeDataTask.ContinueWith(
```

Comment:
This ContinueWIth looks to be not needed, please do just 
batch.Responses = await peer.GetNodeData(hashes, token)

Reply:
will not touch it for now - it was like this, I was already thinking about changing but it may go horribly wrong

Reply:
your solution would not work - await would cause this to be non-parallel

Reply:
wait, it may be a different thing that I think of

Reply:
sorry I thought you meant this one:
                    Task task = ExecuteRequest(token, request);
#pragma warning disable 4014
                    task.ContinueWith(t =>
#pragma warning restore 4014
                    {
                        Interlocked.Decrement(ref _pendingRequests);
                    });

Reply:
but await on continuation swallows the excepotion, while the other one would throw

Reply:
not touching for now - too risky

Reply:
I just mvoed it from other method - it is nothing new